File: mycodes/1)arrays/p13/p13.py

def findsubarray(arr,givensum):
    left = 0
    right = 0
    currrent_sum = 0


    while(currrent_sum != givensum):
#step1)increase right pointer towards right and keep left pointer stable        
        if currrent_sum < givensum:
            currrent_sum = currrent_sum +arr[right]

            if currrent_sum <givensum:
                right = right+1

#step2)increase left pointer towards right and substract current sum till currentsum==given sum        
            
        else:
            currrent_sum = currrent_sum - arr[left]
            left=left+1
           

    print("left is {}".format(left))


    print("right is {}".format(right))

    print("For givensum = {} Subarray is {}".format(givensum,arr[left:right+1]))






if __name__ == '__main__':
    arr = [1,4,20,3,10,5]

    givensum = 33

    findsubarray(arr,givensum)




'''
ref:
https://www.youtube.com/watch?v=Ofl4KgFhLsM

Time complexity = O(n)

Space Complexity = O(1)



Output:


C:\\Users\\Viplav\\Desktop\\WORK\\placements\\python>python demo.py

left is 2
right is 4
For givensum = 33 Subarray is [20, 3, 10]





'''

File: mycodes/1)arrays/p13/p13.c
#include<stdio.h>  
#include<limits.h>
//#Find Contiguous Subarray with given sum


int main()
{
 int arr[] = {1,4,20,3,10,5};

 int sum = 33;

 int left = 0;

 int right = 0;
  
 int size = sizeof(arr)/sizeof(int);
 
 int current_sum = 0;

while(current_sum != sum)
{
         

		  if(current_sum < sum)
		  {
		  	current_sum = current_sum + arr[right];
		  	
		  	if(current_sum < sum)
		  	  {

		  	   right  = right +1;

		  	  }


		  }
		  else
		  {
		   
		  current_sum = current_sum - arr[left];

		  left = left+1;


		  }

}

printf("left : %d , right :%d",left,right);
printf("\nSubarray is:\n");

for(int j = left;j<=right;j++)
{


printf("%d\t",arr[j]);



}



 return 0;
}



/*
https://www.youtube.com/watch?v=Ofl4KgFhLsM&ab_channel=TECHDOSE
Time complexity = O(2*n) =O(n)

Space Complexity = O(1)














Steps of logic/Algo

1)left = 0,right = 0

2)while(currentsum != sum)
 {
	if()
	{
	keep increasing currentsum by arr element from left traversing

	keep increment right pointer

	}
	else{
		keep decreasing currentsum by arr element from left traversing

        keep incrementing left pointer

	}

 }





*/